File: Data Structures & Algorithms/products-of-array-discluding-self/submission-0.kt
class Solution {
    fun productExceptSelf(nums: IntArray): IntArray {

        val result = IntArray(nums.size) { 1 }
        val n = nums.size

        var leftProduct = 1 
        for(i in 0 until n) {
            result[i] = leftProduct
            leftProduct *= nums[i]
        }

        var rightProduct = 1 
        for(i in n-1 downTo 0) {
            result[i] *= rightProduct
            rightProduct *= nums[i]
        }

        return result
    }
}

File: Data Structures & Algorithms/sum-of-all-subset-xor-totals/submission-0.py
class Solution:
    def subsetXORSum(self, nums: List[int]) -> int:
        
        result = 0

        def dfs(i, xorsum):
            nonlocal result
            if i == len(nums):
                result += xorsum
                return
            
            dfs(i + 1, xorsum)
            dfs(i + 1, xorsum ^ nums[i])
        
        dfs(0, 0)
        return result